// @lovable.dev/vite-plugin-dev-server-bridge/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region src/index.ts
const RESTART_ANNOUNCE_PATH = "/__lovable/restart-announce";
const RESTART_EVENT = "lovable:dev-server-restarting";
const PREVIEW_EXECUTE_PATH = "/_sandbox/preview/execute";
const PREVIEW_VIEWERS_PATH = "/_sandbox/preview/viewers";
const PREVIEW_EXECUTE_EVENT = "lovable:preview-execute";
const PREVIEW_EXECUTE_RESULT_EVENT = "lovable:preview-execute-result";
const PREVIEW_EXECUTE_CLIENT_MODULE_ID = "virtual:lovable-preview-execute-client";
const RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID = "\0virtual:lovable-preview-execute-client";
const RESTART_MAX_BODY_BYTES = 1024;
const PREVIEW_EXECUTE_MAX_BODY_BYTES = 64 * 1024;
const PREVIEW_EXECUTE_MAX_CODE_LENGTH = 50 * 1024;
const PREVIEW_EXECUTE_DEFAULT_TIMEOUT_MS = 1500;
const PREVIEW_EXECUTE_MAX_TIMEOUT_MS = 5e3;
let requestSequence = 0;
function nextRequestID() {
	requestSequence += 1;
	return `preview-exec-${Date.now().toString(36)}-${requestSequence.toString(36)}`;
}
function readJsonBody(req, maxBodyBytes) {
	return new Promise((resolve) => {
		let received = 0;
		const chunks = [];
		let aborted = false;
		req.on("data", (chunk) => {
			if (aborted) return;
			received += chunk.length;
			if (received > maxBodyBytes) {
				aborted = true;
				resolve({
					ok: false,
					reason: "too_large"
				});
				return;
			}
			chunks.push(chunk);
		});
		req.on("end", () => {
			if (aborted) return;
			if (chunks.length === 0) {
				resolve({
					ok: false,
					reason: "empty"
				});
				return;
			}
			try {
				resolve({
					ok: true,
					value: JSON.parse(Buffer.concat(chunks).toString("utf8"))
				});
			} catch {
				resolve({
					ok: false,
					reason: "invalid_json"
				});
			}
		});
		req.on("error", () => resolve({
			ok: false,
			reason: "read_error"
		}));
	});
}
function writeJson(res, statusCode, value) {
	res.writeHead(statusCode, { "content-type": "application/json; charset=utf-8" });
	res.end(JSON.stringify(value));
}
function clampTimeoutMs(value) {
	if (typeof value !== "number" || !Number.isFinite(value)) return PREVIEW_EXECUTE_DEFAULT_TIMEOUT_MS;
	return Math.max(1, Math.min(PREVIEW_EXECUTE_MAX_TIMEOUT_MS, Math.trunc(value)));
}
function parsePreviewExecuteRequest(value) {
	if (!value || typeof value !== "object") return "body must be a JSON object";
	const code = value.code;
	if (typeof code !== "string" || code.trim() === "") return "code must be a non-empty string";
	if (code.length > PREVIEW_EXECUTE_MAX_CODE_LENGTH) return `code must be at most ${PREVIEW_EXECUTE_MAX_CODE_LENGTH} characters`;
	return {
		code,
		timeout_ms: clampTimeoutMs(value.timeout_ms)
	};
}
function isPreviewExecuteResult(value) {
	if (!value || typeof value !== "object") return false;
	const result = value;
	return typeof result.request_id === "string" && typeof result.viewer_id === "string" && typeof result.ok === "boolean" && typeof result.duration_ms === "number";
}
function executeInViewers(ws, request, pendingRequests) {
	const requestID = nextRequestID();
	const viewerCount = ws.clients?.size ?? 0;
	const results = /* @__PURE__ */ new Map();
	pendingRequests.set(requestID, results);
	if (viewerCount === 0) {
		pendingRequests.delete(requestID);
		return Promise.resolve({
			request_id: requestID,
			viewer_count: 0,
			responded_count: 0,
			timed_out_count: 0,
			results: []
		});
	}
	ws.send({
		type: "custom",
		event: PREVIEW_EXECUTE_EVENT,
		data: {
			request_id: requestID,
			code: request.code
		}
	});
	return new Promise((resolve) => {
		let done = false;
		let timeout;
		let checkComplete;
		const finish = () => {
			if (done) return;
			done = true;
			clearTimeout(timeout);
			clearInterval(checkComplete);
			pendingRequests.delete(requestID);
			resolve({
				request_id: requestID,
				viewer_count: viewerCount,
				responded_count: results.size,
				timed_out_count: Math.max(0, viewerCount - results.size),
				results: [...results.values()]
			});
		};
		timeout = setTimeout(finish, request.timeout_ms);
		checkComplete = setInterval(() => {
			if (results.size >= viewerCount) finish();
		}, 10);
	});
}
function previewExecuteClientScript() {
	return `
if (import.meta.hot) {
  // Two carriers can instantiate this module in one page, and Vite's stale-listener
  // cleanup in createHotContext can silently drop an earlier instance's listeners —
  // so every instance registers, and requests dedupe through globalThis state.
  // Served untranspiled: keep the syntax at ES2020 (no ??=).
  const viewerID =
    globalThis.__lovablePreviewExecuteViewerID ||
    (globalThis.__lovablePreviewExecuteViewerID =
      globalThis.crypto?.randomUUID?.() ?? Math.random().toString(36).slice(2));
  const handled =
    globalThis.__lovablePreviewExecuteHandled || (globalThis.__lovablePreviewExecuteHandled = new Set());
  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
  const serialize = (value) => {
    if (value === undefined) return { type: "undefined" };
    try {
      return structuredClone(value);
    } catch {
      try {
        return JSON.parse(JSON.stringify(value));
      } catch {
        return String(value);
      }
    }
  };
  import.meta.hot.on("${PREVIEW_EXECUTE_EVENT}", async (payload) => {
    const started = performance.now();
    const requestID = payload && typeof payload.request_id === "string" ? payload.request_id : "";
    const code = payload && typeof payload.code === "string" ? payload.code : "";
    if (!requestID || !code || handled.has(requestID)) return;
    // Request ids embed a timestamp + sequence and are never reused, so clearing
    // at the cap cannot resurrect an old request; it only bounds realm lifetime growth.
    if (handled.size >= 10000) handled.clear();
    handled.add(requestID);
    try {
      const value = await new AsyncFunction(code)();
      import.meta.hot.send("${PREVIEW_EXECUTE_RESULT_EVENT}", {
        request_id: requestID,
        viewer_id: viewerID,
        ok: true,
        value: serialize(value),
        duration_ms: Math.round(performance.now() - started),
      });
    } catch (err) {
      import.meta.hot.send("${PREVIEW_EXECUTE_RESULT_EVENT}", {
        request_id: requestID,
        viewer_id: viewerID,
        ok: false,
        error: {
          name: err && typeof err === "object" && "name" in err ? String(err.name) : undefined,
          message: err instanceof Error ? err.message : String(err),
          stack: err instanceof Error ? err.stack : undefined,
        },
        duration_ms: Math.round(performance.now() - started),
      });
    }
  });
}
`;
}
//   @vitejs/plugin-react and vitejs/devtools use — see vitejs/vite#22864. Gated to
function isPreviewExecuteClientCarrier(id) {
	return id.includes("/vite/dist/client/client.mjs");
}
function devServerBridgePlugin() {
	let isBundledDev = false;
	const injectBundledDevCarrier = () => {
		if (!isBundledDev) return void 0;
		return [{
			tag: "script",
			attrs: { type: "module" },
			children: `import ${JSON.stringify(PREVIEW_EXECUTE_CLIENT_MODULE_ID)};`,
			injectTo: "head"
		}];
	};
	return {
		name: "lovable-dev-server-bridge",
		apply: "serve",
		configResolved(config) {
			isBundledDev = config.experimental?.bundledDev === true;
		},
		configureServer(srv) {
			const pendingPreviewExecuteRequests = /* @__PURE__ */ new Map();
			const ws = srv.ws;
			let connectionSeq = 0;
			ws.on?.("connection", () => {
				connectionSeq += 1;
			});
			ws.on?.(PREVIEW_EXECUTE_RESULT_EVENT, (data, client) => {
				if (!isPreviewExecuteResult(data)) return;
				const results = pendingPreviewExecuteRequests.get(data.request_id);
				if (!results) return;
				results.set(client ?? data.viewer_id, data);
			});
			srv.middlewares.use((req, res, next) => {
				const path = (req.url ?? "").split("?", 1)[0];
				if (req.method === "GET" && path === PREVIEW_VIEWERS_PATH) {
					writeJson(res, 200, {
						viewer_count: ws.clients?.size ?? 0,
						connection_seq: connectionSeq
					});
					return;
				}
				if (req.method !== "POST") return next();
				if (path !== RESTART_ANNOUNCE_PATH && path !== PREVIEW_EXECUTE_PATH) return next();
				readJsonBody(req, path === PREVIEW_EXECUTE_PATH ? PREVIEW_EXECUTE_MAX_BODY_BYTES : RESTART_MAX_BODY_BYTES).then((parsed) => {
					if (path === PREVIEW_EXECUTE_PATH) {
						if (!parsed.ok) {
							writeJson(res, parsed.reason === "too_large" ? 413 : 400, { error: parsed.reason });
							return;
						}
						const previewRequest = parsePreviewExecuteRequest(parsed.value);
						if (typeof previewRequest === "string") {
							writeJson(res, 400, { error: previewRequest });
							return;
						}
						executeInViewers(ws, previewRequest, pendingPreviewExecuteRequests).then((response) => {
							writeJson(res, 200, response);
						});
						return;
					}
					const reason = parsed.ok && parsed.value && typeof parsed.value === "object" && "reason" in parsed.value && typeof parsed.value.reason === "string" ? parsed.value.reason : void 0;
					srv.ws.send({
						type: "custom",
						event: RESTART_EVENT,
						data: reason ? { reason } : {}
					});
					res.writeHead(204);
					res.end();
				});
			});
		},
		resolveId(id) {
			if (id === PREVIEW_EXECUTE_CLIENT_MODULE_ID) return RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID;
			return null;
		},
		load(id) {
			if (id === RESOLVED_PREVIEW_EXECUTE_CLIENT_MODULE_ID) return previewExecuteClientScript();
			return null;
		},
		transform(code, id, options) {
			if (options?.ssr) return null;
			if (!isPreviewExecuteClientCarrier(id)) return null;
			return {
				code: `${code}\nimport(${JSON.stringify(PREVIEW_EXECUTE_CLIENT_MODULE_ID)});\n`,
				map: null
			};
		},
		transformIndexHtml: {
			order: "pre",
			handler: injectBundledDevCarrier,
			enforce: "pre",
			transform: injectBundledDevCarrier
		}
	};
}
//#endregion
exports.devServerBridgePlugin = devServerBridgePlugin;
